### DISCUSSION TITLE
Metadata module versioning and attestation validation

### DISCUSSION THREAD

**Discussion by mpeng-okta** [webauthn4j/webauthn4j#682] [Q&A] [answered]
I had a couple quick questions about the `webauthn4j-metadata` module's naming and versioning, as well as confirming the approach to attestation validation during registration, whenever you had a moment.

First, the documentation calls out the following:
> webauthn4j-meta module, which provides FIDO Metadata Statement handling, is under experimental status.

So to include this module, is it named `webauthn4j-metadata` or `webauthn4j-meta`? Sounds like it's the former, since later there's a reference to `webauthn4j-metadata.jar`, so it's probably just a typo in the guidance text, but I just wanted to double-check.

Also, about this bit farther down:
> The included classes don’t follow semantic versioning and the design may be changed even though it is public.

Does this mean that the `metadata` module is completely decoupled from the `core` module and the rest of the webauthn4j modules, and has no connection with respect to the versioning? So the module might get breaking changes without notice?

Is this also the intention indefinitely going forward -- if not, would there eventually be a separately-versioned repo for the module, or perhaps would this eventually get tied into the same versions? The use case here does involve attestation validation, and my understanding is that the metadata module will be required if we want to validate attestations. Of course, if it's still considered unstable with possible breaking changes, then perhaps we would have to hold off on attestation validation via webauthn4j.

Finally, am I correct in my understanding that, in order to validate the attestation using the blob provided by the FIDO MDS (with a locally-downloaded version of the blob) using webauthn4j, we would do the following:
1. Pass `DefaultCertPathTrustworthinessValidator` into `WebAuthnRegistrationManager`
2. `DefaultCertPathTrustworthinessValidator` needs a `TrustAnchorRepository` -- for this purpose, we would use `MetadataBLOBBasedTrustAnchorRepository`
3. `MetadataBLOBBasedTrustAnchorRepository` needs to build a `MetadataBLOBBasedMetadataStatementRepository` using one or more `MetadataBLOBProvider`s -- for this purpose,  since we're trying to use a local file, we would use an instance of `LocalFileMetadataBLOBProvider`
4. Finally, `LocalFileMetadataBLOBProvider` needs a path to the local file, which in this case would just be the path to the entire `blob.jwt` downloaded from https://mds.fidoalliance.org/ 

Thanks so much in advance for your time and help! This project is awesome.

**Comment by ynojima** [comment, marked as answer]
> webauthn4j-meta module, which provides FIDO Metadata Statement handling, is under experimental status.

Sorry, this is a typo. Thank you for pointing out.

> Does this mean that the metadata module is completely decoupled from the core module and the rest of the webauthn4j modules, and has no connection with respect to the versioning? So the module might get breaking changes without notice?

The sentence here has some background.

Originally, I meant to announce WebAuthn4J version 1.0 a year or so ago.
At that time, Fido MDS was version 2, and the MDS specifications did not seem stable.
I thought I need to allow breaking change in the minor release of `webauthn4j-metadata` module even after declaring WebAuthn4J version 1.0.
That's why I marked `webauthn4j-metadata` experimental.

But after that, MDS3 was released, WebAuthn Level 3 spec is compiling, I still have no time to declare WebAuthn4J 1.0.
so the sentence is not a very meaningful for now.
I bump minor version if it has breaking change or big new feature.
If not, I bump patch version.

**Reply by ynojima** [reply]
> Finally, am I correct in my understanding that, in order to validate the attestation using the blob provided by the FIDO MDS (with a locally-downloaded version of the blob) using webauthn4j, we would do the following:

Correct.

**Reply by mpeng-okta** [reply]
Yeah, it made sense to declare the module experimental at that time, since there's that gap between v2 and v3 of the MDS. And yeah, I can see why 1.0 wouldn't declared quite yet, since the WebAuthn level 3 spec is imminent.

Cool, so in that sense, I suppose the metadata module as it is currently is no longer experimental, and it would be reasonable to consider it as being on the same overall webauthn4j version as the core module?

The versioning info is useful as well; my understanding was that the major version would indicate breaking changes and that minor version bumps would indicate backward compatible functionality changes. But it appears this is not the case -- minor version upgrade can also indicate breaking changes, so it's something for which I should watch out. Got it, thanks!

**Reply by mpeng-okta** [reply]
Hm, one more follow-up. Locally, we already have a json blob filtered out to use only FIDO2 objects from the MDS blob. This exists as a JSON file, which is a list of each metadata statement for each WebAuthn model.

My question (hopefully not too silly) is: would there be a natural way of integrating this as a `TrustAnchorRepository` via WebAuthn4J, or will we end up just having to use the entire `blob.jwt` from the FIDO MDS if we want to integrate this properly with WebAuthn4J?

**Reply by ynojima** [reply]
> Cool, so in that sense, I suppose the metadata module as it is currently is no longer experimental, and it would be reasonable to consider it as being on the same overall webauthn4j version as the core module?

There is still some gap between the maturity level of `webauthn4j-core` and `webauthn4j-metadata`, but I think `webauthn4j-metadata` is close to graduate `experimental`.

> major version would indicate breaking changes and that minor version bumps would indicate backward compatible functionality changes.

This is only true when the majar version > 0.
https://semver.org/#spec-item-8

> minor version upgrade can also indicate breaking changes, so it's something for which I should watch out.

As long as webauthn4j hasn't declare v1.0, yes.

> Locally, we already have a json blob filtered out to use only FIDO2 objects from the MDS blob. This exists as a JSON file, which is a list of each metadata statement for each WebAuthn model.

If you have metadata statement JSON files instead of a metadata BLOB JWT file,
You can use `MetadataStatementsBasedTrustAnchorRepository` instead of `MetadataBLOBBasedTrustAnchorRepository` , and inject `MetadataStatementsProvider` implementation (`LocalFilesMetadataStatementsProvider` or your own) to it.

**Reply by mpeng-okta** [reply]
> This is only true when the majar version > 0.

Ooh, I didn't know that. I just realized item 4 also covers this (i.e. pre major version, anything may change at any time). Hm, OK. Yeah, definitely might behoove me to be cautious about integrating anything for the time being, then. Thanks for educating me!

I totally understand that you're essentially maintaining this solo, on your own time, and may not have any concrete estimates; so, feel free to say you have no idea for the time being. I was wondering though, would you have a very rough idea for when you might plan on releasing the first major version?

> If you have metadata statement JSON files instead of a metadata BLOB JWT file,
You can use `MetadataStatementsBasedTrustAnchorRepository` instead of `MetadataBLOBBasedTrustAnchorRepository` , and inject `MetadataStatementsProvider` implementation (`LocalFilesMetadataStatementsProvider` or your own) to it.

Ah nice, I did take a look at that, but it clearly slipped past me. I guess my file is one single JSON-format list of metadata objects, so I'd have to reimplement `MetadataStatementsProvider` to accept this format (vs individual objects), but I see what you mean.

(edited since I had a couple misunderstandings initially) I did realize that the framework doesn't have a caching provider for local metadata statement (only when it comes to blob provider). So yeah, that's a consideration of which I'll be aware.

Thanks so much for your time and help!

**Reply by ynojima** [reply]
> Yeah, definitely might behoove me to be cautious about integrating anything for the time being

As WebAuthn4J is used by Keycloak and Red Hat Single Sign On, I will try to avoid breaking changes as long as possible, but if it is inevitable, I will explicitly write it in the release note.

> would you have a very rough idea for when you might plan on releasing the first major version?

That's the question. I want to implement passkey support, and want to confirm webauthn4j works well with major browsers' passkey implementation before releasing v1.0.0.

> I'd have to reimplement MetadataStatementsProvider to accept this format

Yes, if you need to support original format, please create your own `MetadataStatementsProvider` implementation, and you can implement caching mechanism there as you like.

**Reply by mpeng-okta** [reply]
That all makes sense. Thank you so much!
